src/qibo/hamiltonians/hamiltonians.py

```diff
@@ -663,120 +621,40 @@ def expectation_from_samples(self, freq: dict, qubit_map: list = None) -> float:
-            if self._terms is not None and o._terms is not None:
-                new_ham.terms = self.terms + o.terms
-                new_ham.constant = self.constant + o.constant
-            if self._dense is not None and o._dense is not None:
-                new_ham.dense = self.dense + o.dense
 
-        elif isinstance(o, self.backend.numeric_types):
-            new_ham = self.__class__(
-                symbol_map=dict(self.symbol_map), backend=self.backend
-            )
-            if self._form is not None:
-                new_ham.form = self.form + o
-            if self._terms is not None:
-                new_ham.terms = self.terms
-                new_ham.constant = self.constant + o
-            if self._dense is not None:
-                new_ham.dense = self.dense + o
+            if o._form is not None:
+                form = operator(form, o._form) if form is not None else o._form
 
+        elif isinstance(o, (self.backend.numeric_types, self.backend.tensor_types)):
+            form = operator(form, complex(o)) if form is not None else complex(o)
         else:
             raise_error(
                 NotImplementedError,
-                f"SymbolicHamiltonian addition to {type(o)} not implemented.",
+                f"SymbolicHamiltonian composition to {type(o)} not implemented.",
             )
-        return new_ham
 
-    def __sub__(self, o):
-        if isinstance(o, self.__class__):
-            if self.nqubits != o.nqubits:
-                raise_error(
-                    RuntimeError,
-                    "Only hamiltonians with the same number of qubits can be subtracted.",
-                )
-            new_ham = self.__class__(
-                symbol_map=dict(self.symbol_map), backend=self.backend
-            )
-            if self._form is not None and o._form is not None:
-                new_ham.form = self.form - o.form
-                new_ham.symbol_map.update(o.symbol_map)
-            if self._terms is not None and o._terms is not None:
-                new_ham.terms = self.terms + [-1 * x for x in o.terms]
-                new_ham.constant = self.constant - o.constant
-            if self._dense is not None and o._dense is not None:
-                new_ham.dense = self.dense - o.dense
+        return self.__class__(form=form, nqubits=self.nqubits, backend=self.backend)
 
-        elif isinstance(o, self.backend.numeric_types):
-            new_ham = self.__class__(
-                symbol_map=dict(self.symbol_map), backend=self.backend
-            )
-            if self._form is not None:
-                new_ham.form = self.form - o
-            if self._terms is not None:
-                new_ham.terms = self.terms
-                new_ham.constant = self.constant - o
-            if self._dense is not None:
-                new_ham.dense = self.dense - o
+    def __add__(self, o):
+        return self._compose(o, lambda x, y: x + y)
```

Comment:
Here (and in the other similar function) you can also use standard library's [`operator`](https://docs.python.org/3/library/operator.html)
```suggestion
        import operator
        
        return self._compose(o, operator.add)
```